### PR TITLE
Add apache http client

### LINKED ISSUE #1139: Support configuring the socket options.
### Usage:

In my case I have 2 zones one for computation with Spark the other for result storage with ClickHouse and the bond width between the 2 zones is limited.

I have 2 task with different priorities and I want the hight priority task can transfer data without the effecting of the low priority task.

### Possible solution:

Support configuring the ip dscp. 

But `clickhouse-jdbc`  use `java.net.HttpURLConnection` which does support set socket options. 

Can somebody help?  /cc @zhicwu

### PR DESCRIPTION
**Pull request by JackyWoo**
This pr will close #1139 .

### change log
1. Add apache http client
2. who support configuring socket option.
3. Make `ClickHouseHttpClientTest` parameterized

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -99,6 +99,7 @@
         <testcontainers.version>1.17.5</testcontainers.version>
         <!-- stay 7.5 for JDK 8 support -->
         <testng.version>7.5</testng.version>
+        <apache.httpclient.version>4.5.13</apache.httpclient.version>
```
Comment (zhicwu) [inline pom.xml:102]: The legacy driver was using http client 4.5 and it's known to have "no response" error sometimes. Can we [migrate to 5.2](https://hc.apache.org/httpcomponents-client-5.2.x/migration-guide/preparation.html)?

File: clickhouse-client/src/test/java/com/clickhouse/client/ClientIntegrationTest.java [hunk 2]
```diff
@@ -37,6 +39,7 @@
 import com.clickhouse.client.ClickHouseTransaction.XID;
 import com.clickhouse.client.config.ClickHouseBufferingMode;
 import com.clickhouse.client.config.ClickHouseClientOption;
+import com.clickhouse.client.config.ClickHouseOption;
```
Comment (zhicwu) [inline clickhouse-client/src/test/java/com/clickhouse/client/ClientIntegrationTest.java:42]: Remove unnecessary changes.

File: clickhouse-client/src/main/java/com/clickhouse/client/config/ClickHouseSocketOption.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Socket related options.
+ */
+public enum ClickHouseSocketOption implements ClickHouseOption {
+
+    /**
+     * Socket IP_TOS option which indicates IP package priority.
+     */
+    IP_TOS("socket_op_ip_tos", 0, "Socket IP_TOS option which indicates IP package priority.");
```
Comment (zhicwu) [inline clickhouse-client/src/main/java/com/clickhouse/client/config/ClickHouseSocketOption.java:16]: Why not simply add this into `ClickHouseClientOption`?

**Review by zhicwu — CHANGES_REQUESTED** [review]

**Comment by zhicwu** [comment]
Thank you @JackyWoo for the pull request. Just added a few immediate feedback. Apart from that, we also need to update pom to ensure Apache libraries will be only included in -all jar. Will run benchmark tomorrow to understand more.

**Comment by zhicwu** [comment]
The performance does not look good somehow.

```
Benchmark                                     (client)  (connection)  (statement)   (type)   Mode  Cnt    Score   Error  Units
Query.selectInt8    clickhouse-apache-http-client-jdbc         reuse     prepared  default  thrpt   20   95.181 ± 3.783  ops/s
Query.selectInt8                  clickhouse-http-jdbc         reuse     prepared  default  thrpt   20  112.334 ± 7.949  ops/s
Query.selectString  clickhouse-apache-http-client-jdbc         reuse     prepared  default  thrpt   20   27.539 ± 0.925  ops/s
Query.selectString                clickhouse-http-jdbc         reuse     prepared  default  thrpt   20   35.449 ± 2.106  ops/s
Query.selectUInt64  clickhouse-apache-http-client-jdbc         reuse     prepared  default  thrpt   20   45.539 ± 1.620  ops/s
Query.selectUInt64                clickhouse-http-jdbc         reuse     prepared  default  thrpt   20   62.749 ± 4.282  ops/s
```

**Comment by JackyWoo** [comment]
May I ask how to configure `client`?

**Comment by zhicwu** [comment]
> May I ask how to configure benchmark `client` to `clickhouse-apache-http-client-jdbc`?

Please pull latest code fron your branch and follow instructions below.

```bash
cd clickhouse-benchmark
mvn -Drelease clean package
java -jar target/benchmarks.jar -p client=clickhouse-apache-http-client-jdbc -p type=default -p connection=reuse Query.selectInt8
```

**Comment by JackyWoo** [comment]
Optimize apache http client by reuse connection. Below is benchmark in my local machine.

````
Benchmark               (client)  (connection)  (statement)   (type)   Mode  Cnt    Score   Error  Units
Query.selectInt8  ClickhouseJdbc         reuse     prepared  default  thrpt   20   96.870 ± 5.932  ops/s
Query.selectInt8  ClickhouseJdbc         reuse     prepared   string  thrpt   20   96.102 ± 7.897  ops/s
Query.selectInt8  ClickhouseJdbc         reuse     prepared   object  thrpt   20  100.323 ± 8.398  ops/s

Benchmark                               (client)  (connection)  (statement)   (type)   Mode  Cnt    Score   Error  Units
Query.selectInt8  ClickhouseApacheHttpClientJdbc         reuse     prepared  default  thrpt   20  107.038 ± 4.379  ops/s
Query.selectInt8  ClickhouseApacheHttpClientJdbc         reuse     prepared   string  thrpt   20  107.169 ± 3.687  ops/s
Query.selectInt8  ClickhouseApacheHttpClientJdbc         reuse     prepared   object  thrpt   20  109.726 ± 2.098  ops/s

```

**Comment by JackyWoo** [comment]
@zhicwu Please review the PR, when you have a changce.

**Comment by zhicwu** [comment]
Thank you for the contribution @JackyWoo. I was waiting the build failure got fixed :)

The benchmark makes sense as they should be similar, and the result could be slightly different when using single thread vs. multi-thread for the benchmark.

Anyway, just document my concerns and may change the code after merging this PR.
* Is `ByteArraysInputStream` really needed?
* Consider `CloseableHttpAsyncClient`
* Packing - change httpclient5 to provided?
* Stress test to understand if we still have "no response" error

Lastly, while I'm working on TCP client, if we need better control of Socket, we probably need non-blocking SocketChannel based implementation as well.

**Review by zhicwu — APPROVED** [review]

**Comment by JackyWoo** [comment]
@zhicwu Thank you,  may I know when 0.33 will release?

**Pull request merged**